dlrover/python/common/event/context.py

```diff
@@ -77,80 +96,115 @@ def add_step_event(self, event: AtorchEvent):
-                    event.name, TrainEventState.TRAIN_EVT_ONCE, event.step
-                )
-                step_event.begin_timestamp = event.timestamp
-                step_event.end_timestamp = event.timestamp
-                self._step_events[event.timestamp] = step_event
-
     def add_ckpt_event(self, event: AtorchEvent):
         with self._lock:
             keys = list(self._step_events.keys())
-            if not len(keys):
-                logger.error("empty step events")
-                return
-            elif event.timestamp is None or event.step is None:
+            if event.timestamp is None or event.step is None:
                 logger.error(f"invalid event: {event}")
                 return
+            elif len(keys) >= self._max_events:
+                self._step_events.popitem(last=False)
 
-            last_key = keys[-1]
-            last_step = self._step_events[last_key]
-            if last_step.end_timestamp is None:
-                logger.error(f"invalid last step: {last_step}")
-                return
-            if (
-                event.step != last_step.step
-                or event.timestamp < last_step.end_timestamp
-            ):
-                logger.error(f"invalid last step: {last_step}, {event}")
-                return
+            if event.type == EventTypeName.BEGIN:
+                if len(keys) > 0:
+                    last_event = self._step_events[keys[-1]]
+                    if last_event.event_state != TrainEventState.TRAIN_EVT_END:
+                        logger.error(
+                            f"invalid ckpt step: {last_event}, {event}"
+                        )
+                        return
+                    if event.timestamp < last_event.end_timestamp:
+                        logger.error(
+                            f"invalid ckpt step time: {last_event}, {event}"
+                        )
+                        return
+
+                step_event = AtorchStepEvent(
+                    event.name, TrainEventState.TRAIN_EVT_BEGIN, event.step
+                )
+                step_event.begin_timestamp = event.timestamp
+                step_event.localtime = int(datetime.now().timestamp())
+                self._step_events[event.timestamp] = step_event
+
+            elif event.type == EventTypeName.END:
+                if not len(keys):
+                    logger.error(f"invalid ckpt step without BEGIN: {event}")
+                    return
+                last_key = keys[-1]
+                step_event = self._step_events[last_key]
+                if (
+                    step_event.step != event.step
+                    or step_event.event_state
+                    != TrainEventState.TRAIN_EVT_BEGIN
+                ):
+                    logger.error(f"invalid ckpt step: {step_event}, {event}")
+                    return
+                if step_event.begin_timestamp > event.timestamp:
+                    logger.error(
```

Comment:
There are too many error logs of this type. There should be a unified event validation process here, where validation failures are logged as warnings. 
BTW, currently only event logs that impact core processes use the error level.